internal/daemon/api_files_test.go

```diff
@@ -1054,21 +1054,33 @@ func readMultipart(c *C, response *http.Response, body io.Reader, result interfa
 	c.Assert(mediaType, Equals, "multipart/form-data")
 	c.Assert(params["boundary"], Not(Equals), "")
 	mr := multipart.NewReader(body, params["boundary"])
-	form, err := mr.ReadForm(4096)
-	c.Assert(err, IsNil)
-
-	err = json.Unmarshal([]byte(form.Value["response"][0]), result)
-	c.Assert(err, IsNil)
 
+	// First decode all the files
 	files := make(map[string]string)
-	for _, fh := range form.File["files"] {
-		f, err := fh.Open()
+	var part *multipart.Part
+	for {
+		part, err = mr.NextPart()
 		c.Assert(err, IsNil)
-		b, err := ioutil.ReadAll(f)
+		if part.FormName() != "files" {
```

Comment:
This seems a significant difference in the looping. In the former we were iterating over everything in 'files' now we seem to be iterating just over the multipart but then not iterating over files.

(The old form looked like the data was something like:
```
{'blah': ?,
'files': [content, content, content],
}
```

But the new way we are treating it seems more like:
```['blah', {'files': content}, {'files': content'}, {'files': content}]```

I'm assuming that the test suite is passing so I'm wrong about something (I seem to recall some Python HTTP classes that would let you treat repeated headers as a single list.)

Reply:
Yeah, I think you're assuming something wrong. The raw read-files response (before and after) is like so:

```
--01234567890123456789012345678901
Content-Disposition: form-data; name="files"; filename="/root/hi.txt"

Hello
--01234567890123456789012345678901
Content-Disposition: form-data; name="files"; filename="/root/bye.txt"

Bye bye
--01234567890123456789012345678901
Content-Disposition: form-data; name="files"; filename="/root/foobar.txt"

Foo
Bar
--01234567890123456789012345678901
Content-Disposition: form-data; name="response"
{
    "action": "read",
    "files": [
        {"path": "/root/hi.txt"},
        {"path": "/root/bye.txt"},
        {"path": "/root/foobar.txt"}
    ]
}
--01234567890123456789012345678901--
```

In the old code, we called `multipart.Reader.ReadForm` to loop over the multipart sections (files and response). In the new code, we do that loop ourselves and don't call `ReadForm` at all -- either way it loops over the files first, and then grabs the response at the end. Hope that helps.